README.md
# DuetGPT

**🎉 News:** DuetGPT no longer uses langchain but instead [OpenAI functions](https://openai.com/blog/function-calling-and-other-api-updates). This means a significant improvement in reliability and performance.

DuetGPT is an experimental AI powered CLI tool and semi-autonomous agent that helps developers with coding tasks and file system tasks. The developer describes tasks to the AI who then issues commands or follow-up questions for clarification. After approval by the developer, **DuetGPT automatically executes the commands issued by the AI.**

DuetGPT also does really well as a general bash helper.

Works with OpenAI models:

- gpt-3.5-turbo-0613 (does not produce any great code though)
- gpt-4-0613

Example tasks:

- Refactor index.js: add inline comments, improve variable naming.
- Write a bash script that lists all cars in the database.
- Find all files in current directory and subdirectories that contain the word "DuetGPT"
- Write a PR description based on the commit messages in the PR.

DuetGPT builds on OpenAI's GPT-4 language model and uses its conversational capabilities to engage in a two-way conversation with the developer. It uses the newly released feature of the OpenAI API that allows the AI to make [function calls](https://openai.com/blog/function-calling-and-other-api-updates).

### ⚠️ DuetGPT has no guardrails! Make sure you understand the commands given by the AI before executing them. ⚠️

Here is a 30 second demo, use DuetGPT to build a Node.js app that draws a mandelbrot fractal using ASCII characters to the console:

https://github.com/kristoferlund/duet-gpt/assets/9698363/147bdf9d-28df-4421-9297-05abfeb53142

## Setup

Install the DuetGPT CLI tool globally using npm:

```bash
npm install -g duet-gpt
```

## Usage

To start DuetGPT, run the following command:

```bash
duet-gpt
```

You will be prompted to enter your OpenAI API key. You can find your API key on the [OpenAI dashboard](https://platform.openai.com/account/api-keys).

```bash
     _            _              _
    | |          | |            | |
  __| |_   _  ___| |_ __ _ _ __ | |_
 / _` | | | |/ _ \ __/ _` | '_ \| __|
| (_| | |_| |  __/ || (_| | |_) | |_
 \__,_|\__,_|\___|\__\__, | .__/ \__|
                      __/ | |
                     |___/|_|

┌  DuetGPT
│
◆  Please enter your OpenAI API key:
│  sk-XXX…
└
```

## Setup for development

1. Clone the repository.

2. Install dependencies:

```bash
npm install
```

3. Run DuetGPT using yarn:

```bash
npm start
```

## Sample interaction

For the purposes of illustrating how interactions work, I chose a super simple task. Let's say you want to add `.yarn/cache` to `.gitignore`. Here is how to ask DuetGPT to help you with that:

```
     _            _              _
    | |          | |            | |
  __| |_   _  ___| |_ __ _ _ __ | |_
 / _` | | | |/ _ \ __/ _` | '_ \| __|
| (_| | |_| |  __/ || (_| | |_) | |_
 \__,_|\__,_|\___|\__\__, | .__/ \__|
                      __/ | |
                     |___/|_|

┌  DuetGPT
│
◇  LLM and memory started
│
┌ 🤖 ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                                                                                            │
│                                                                                                                            │
│    My mission is to collaborate with you as a developer by providing specific bash or file modification commands to help   │
│    you complete your development-related tasks, while effectively engaging in two-way interactions.                        │
│                                                                                                                            │
│                                                                                                                            │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
│
◇  What would you like to do?
│  Add .yarn/cache to .gitignore
│
◇  AI response received
│
┌ 🤖 ────────────────────────────────────────────────────────────────────────┐
│                                                                            │
│                                                                            │
│    First, I need to check if .gitignore exists in the current directory.   │
│                                                                            │
│     § BASH                                                                 │
│                                                                            │
│     1 ls .gitignore                                                        │
│                                                                            │
│                                                                            │
└────────────────────────────────────────────────────────────────────────────┘
│
◇  What would you like to do?
│  Run command
│
┌ 💻 ────────────┐
│                │
│   .gitignore   │
│                │
└────────────────┘
│
◇  AI response received
│
┌ 🤖 ────────────────────────────────────────────────────────────────────┐
│                                                                        │
│                                                                        │
│    I will now append '.yarn/cache' to your existing .gitignore file.   │
│                                                                        │
│     § BASH                                                             │
│                                                                        │
│     1 echo ".yarn/cache" >> .gitignore                                 │
│                                                                        │
│                                                                        │
└────────────────────────────────────────────────────────────────────────┘
│
◇  What would you like to do?
│  Run command
│
┌ 💻 ────┐
│        │
│   OK   │
│        │
└────────┘
│
◆  What would you like to do?
│  Not sure
└

```

## Known issues

- When proposing changes to large files, the AI may return incomplete results. The gpt-4 context window is limited, DuetGPT works best with small files - one function per file is ideal.

## Contributing

Contributions are welcome! Please open an issue or submit a pull request.

## Author

- [kristofer@fmckl.se](mailto:kristofer@fmckl.se)
- Twitter: [@kristoferlund](https://twitter.com/kristoferlund)
- Discord: kristofer#1475

## License

[MIT](LICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.4] - 2023-06-19

### Added

- Added a configuration screen before app starts that allows user to change config (e.g. API key) (7b12610)

## [0.1.2] - 2023-06-19

### Changed

- Notify AI also when function output is empty (1e3ff91)
- Add AI assistant responses to memory (4784b67)
- Throw error in function arguments are malformed (fbb0ed1)
- Make sure spinner stops after error (9b01d91)
- Make sure system.prompt is included in npm package (fd26535)
- Try to get gpt to be more verbose when making function calls (6699913)
- Update README.md (2e35934)
- Merge branch 'fix/minor-issues-230619' of https://github.com/kristoferlund/duet-gpt into fix/minor-issues-230619 (b917603)
- Remove unused (b0356db)

## [0.1.1] - 2022-06-18

### Changed

- Text update (5b2926a, 4fc7a59)
- Removed langchain, use plain OpenAI instead (4e1f913)
- Asked DuetGPT to move welcome messages to separate file (3329f84)

## [0.0.9] - 2022-05-25

### Added

- Initial release of the project.
- Experimental AI developer assistant CLI.
- Developers can give the AI tasks, and the AI responds with commands that the CLI performs.


src/config.ts
import * as fs from "fs";
import * as p from "@clack/prompts";

import os from "os";
import path from "path";

type ConfigPartial = {
  OPEN_AI_KEY?: string;
  OPEN_AI_MODEL?: Model;
};

export type Config = {
  OPEN_AI_KEY: string;
  OPEN_AI_MODEL: Model;
};

export type Model = {
  id: string;
  maxTokens: number;
};

function isCompleteConfig(config: ConfigPartial): config is Config {
  return config.OPEN_AI_KEY !== undefined && config.OPEN_AI_MODEL !== undefined;
}

// Define file and directory names
const configPath = path.join(os.homedir(), ".duet-gpt");

/**
 * Load config from config file
 */
export async function loadConfig(): Promise<ConfigPartial> {
  let config: ConfigPartial = {};
  if (fs.existsSync(configPath)) {
    try {
      config = JSON.parse(fs.readFileSync(configPath, "utf8"));
    } catch (e) {
      throw new Error(`Error parsing config file: ${e}`);
    }
  }
  return config;
}

/**
 * Prompt user for config values
 */
export async function configPrompts(): Promise<Config> {
  let config = await loadConfig();

  // Prompt user for API key
  const apiKey = await p.text({
    message: "Please enter your OpenAI API key: ",
    placeholder: "sk-XXX…",
    initialValue: config.OPEN_AI_KEY,
  });

  if (p.isCancel(apiKey) || !apiKey) {
    p.cancel("OpenAI API key is required to run DuetGPT");
    process.exit(0);
  }

  const model = await p.select<{ value: Model; label: string }[], Model>({
    message: "Select the model to use:",
    initialValue: config.OPEN_AI_MODEL,
    options: [
      {
        value: { id: "gpt-3.5-turbo-0613", maxTokens: 4000 },
        label: "gpt-3.5-turbo-0613",
      },
      { value: { id: "gpt-4-0613", maxTokens: 8000 }, label: "gpt-4-0613" },
    ],
  });

  if (p.isCancel(model) || !model) {
    p.cancel("OpenAI model is required to run DuetGPT");
    process.exit(0);
  }

  // Save API key to config file
  fs.writeFileSync(
    configPath,
    JSON.stringify({ OPEN_AI_KEY: apiKey, OPEN_AI_MODEL: model })
  );

  return { OPEN_AI_KEY: apiKey, OPEN_AI_MODEL: model };
}

/**
 * Prompt user for config or start DuetGPT
 */
export async function configOrStart(): Promise<Config> {
  let config = await loadConfig();

  if (!isCompleteConfig(config)) {
    // If config is not complete, prompt user for to complete it
    config = await configPrompts();
  } else {
    // If config is complete, prompt user to start or change config
    const bootChoice = await p.select({
      message: "Select an action:",
      options: [
        { value: "start", label: "Start" },
        { value: "config", label: "Config" },
      ],
    });

    if (p.isCancel(bootChoice) || !bootChoice) {
      p.cancel("Quitting.. see you next time!");
      process.exit(0);
    }

    // If user chooses to change config, prompt them for new values
    if (bootChoice === "config") {
      config = await configPrompts();
    }
  }

  return config as Config;
}
